Compiler/BackEnd/CommonSubExpression.mo

```diff
@@ -94,105 +97,199 @@ algorithm
+
+  HashSet.HashSet globalKnownVarHT;
 
   DAE.Exp cse, call;
   list<Integer> dependencies;
 algorithm
   size := BackendDAEUtil.maxSizeOfEqSystems(inDAE.eqs) + 42;   //create data structures independent from the size of the EqSystem
   exarray := ExpandableArray.new(size, dummy_equation);
+  exarray_initial := ExpandableArray.new(size, dummy_equation);
 
   size := Util.nextPrime(realInt(2.4*size));
   HT := HashTableExpToIndex.emptyHashTableSized(size);
+  HT_initial := HashTableExpToIndex.emptyHashTableSized(size);
+
+  shared := inDAE.shared;
+  BackendDAE.SHARED(globalKnownVars=globalKnownVars,functionTree=functionTree) := shared;
+
+  // Create Hashtable and store globally known variables in it
+  globalKnownVarHT := HashSet.emptyHashSetSized(Util.nextPrime(realInt(2.4*(globalKnownVars.numberOfVars + 42))));
+  if isSimulationDAE then
+    globalKnownVarHT := BackendVariable.traverseBackendDAEVars(globalKnownVars, VarToGlobalKnownVarHT, globalKnownVarHT);
+  end if;
 
-  BackendDAE.SHARED(functionTree=functionTree) := inDAE.shared;
+  if Flags.isSet(Flags.DUMP_CSE_VERBOSE) then
+    print("Start optimization module wrapFunctionCalls for " + daeTypeStr + " DAE\n" + BORDER + BORDER + "\n\n\n");
+    print("Pre-Phase: Set up data structure\n" + BORDER + "\n");
+    BackendDump.dumpVariables(globalKnownVars, "globalKnownVars before WFC");
+    print("globalKnownVarHT before analysis\n" + UNDERLINE + "\n");
+    BaseHashSet.dumpHashSet(globalKnownVarHT);
+    print("\nPhase 0: Analyse the globalKnownVars\n" + BORDER + "\n");
+  end if;
+
+  if isSimulationDAE then
+    // Traverse globalKnownVars and perform the WFC analysis phase on it
+    (HT_initial, exarray_initial, cseIndex, index_initial, functionTree, globalKnownVarHT) := BackendVariable.traverseBackendDAEVars(globalKnownVars, findCallsInGlobalKnownVars, (HT_initial, exarray_initial, cseIndex, index_initial, functionTree, globalKnownVarHT));
+  end if;
 
+  if Flags.isSet(Flags.DUMP_CSE_VERBOSE) then
+    print("Hastable after analysis of globalKnownVars (HT_initial)\n" + UNDERLINE + "\n");
+    BaseHashTable.dumpHashTable(HT_initial);
+    ExpandableArray.dump(exarray_initial, "\nExpandable Array after analysis of globalKnownVars (exarray_initial)", printCSEEquation);
+    print("\nglobalKnownVarHT after analysis\n" + UNDERLINE + "\n");
+    BaseHashSet.dumpHashSet(globalKnownVarHT);
+  end if;
+
+  // Start the WFC algorithm for all equation systems
   for syst in inDAE.eqs loop
-    HT := BaseHashTable.clear(HT);
-    exarray := ExpandableArray.clear(exarray);
+    if Flags.isSet(Flags.DUMP_CSE_VERBOSE) then
+      print("\n\nHandle system (belongs to " + daeTypeStr + " DAE):\n" + BORDER + "\n");
+      BackendDump.dumpVariables(syst.orderedVars, "Variables");
+      BackendDump.dumpEquationArray(syst.orderedEqs, "Equations");
+      print("\n\nPhase 1: Analysis\n" + BORDER + "\n");
+    end if;
+    HT := BaseHashTable.copy(HT_initial);
```

Comment:
Yes,  a hashtable is always expensive to create because you typically create an array of size relative to all equations: O(n). And here you copy it for each system of equations (also O(n), though often small). O(n^2) is bad though, and if copying lookup tables is necessary, an immutable structure like a trie or tree which has a constant-time copy (just a pointer copy) is preferred.